src/julia_gc.c

```diff
@@ -387,6 +388,36 @@ void MarkJuliaObj(void * obj)
 // from which freed objects are then marked. Hence, we add additional checks
 // when traversing GAP master pointer and bag objects that this happens
 // only for live objects.
+//
+// All stacks in Julia are associated with tasks and we can use
+// jl_task_stack_buffer() to retrieve the buffer information for that stack.
+// That said, in a couple of cases these need adjustments.
+//
+// 1. The stack buffer of the root task of the main thread, when started from
+//    GAP can extend past the point where Julia believes it ends. Therefore,
+//    for that stack, we use GapBottomStack instead.
+// 2. If it is the current task of the current thread, we know where exactly
+//    the lowest valid address of the stack is and do not need to scan the
+//    entire stack buffer.
+//
+// In addition, stack buffers can include guard pages, which trigger a
+// segmentation fault when accessed. We intercept such segmentation faults
+// and terminate stack scanning early for those.
+//
+// In GAP terminology, we use "bottom" to describe the end of the stack
+// where it starts and "top" to describe the current stack pointer. This
+// means that the "bottom" of the stack is the highest address on most
+// modern architectures and "top" is the lowest. When we treat the stack
+// as an abstract buffer, "start" and "end" refer to its lowest and
+// highest address, respectively.
```

Comment:
That's not just "GAP terminology". I found this use of top and bottom in virtually all CS references I checked. Also, I'd bring these definitions earlier, then the text before here could use them instead of talking about origin/start/end/lowest address etc.

Also, I think "the end ... where it starts" is a really confusing choice of words. Perhaps: This:

```suggestion
// We use "bottom" to refer to the origin of the stack, and "top" to describe
// the current stack pointer. Confusingly, on most contemporary architectures,
// the stack grows "downwards", which means that the "bottom" of the stack
// is the highest address and "top" is the lowest. When we treat the stack
// as an abstract buffer, "start" and "end" refer to its lowest and
// highest address, respectively.
```
Also, perhaps instead of "start" and "end" it would then be better to talk about "low" and "high" ?

Reply:
It's conventional terminology when dealing with stacks as abstract data types; when dealing with hardware stacks, terminology is unfortunately inconsistent. Julia in particular uses the opposite terminology for its task stacks ([example](https://github.com/JuliaLang/julia/blob/6cd329c371c1db3d9876bc337e82e274e50420e8/src/task.c#L1058)). This is why I tried to distinguish between Julia and GAP terminology.

However, I think I still like your wording better, as it's overall clearer and more to the point, so I'm good with the change.